storage-proofs/src/hasher/hybrid.rs

```diff
@@ -0,0 +1,256 @@
+use std::cmp::Ordering;
+
+use merkletree::merkle::Element;
+use paired::bls12_381::{Fr, FrRepr};
+use rand::{Rand, Rng};
+use serde::de::DeserializeOwned;
+use serde::ser::Serialize;
+
+use crate::error::Result;
+use crate::hasher::Domain;
+
+#[derive(Clone, Copy, Debug, Eq, Serialize, Deserialize)]
+pub enum HybridDomain<AD, BD>
+where
+    AD: Domain,
+    BD: Domain,
+{
+    #[serde(bound(serialize = "AD: Serialize", deserialize = "AD: DeserializeOwned"))]
+    Alpha(AD),
+
+    #[serde(bound(serialize = "BD: Serialize", deserialize = "BD: DeserializeOwned"))]
+    Beta(BD),
+}
+
+impl<AD, BD> Default for HybridDomain<AD, BD>
+where
+    AD: Domain,
+    BD: Domain,
+{
+    fn default() -> Self {
+        HybridDomain::Alpha(AD::default())
+    }
+}
+
+impl<AD, BD> PartialEq for HybridDomain<AD, BD>
+where
+    AD: Domain,
+    BD: Domain,
+{
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (HybridDomain::Alpha(a), HybridDomain::Alpha(b)) => a == b,
+            (HybridDomain::Beta(a), HybridDomain::Beta(b)) => a == b,
+            _ => panic!("can't compare different variants of `HybridDomain`"),
+        }
+    }
+}
+
+impl<AD, BD> PartialOrd for HybridDomain<AD, BD>
+where
+    AD: Domain,
+    BD: Domain,
+{
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        match (self, other) {
+            (HybridDomain::Alpha(a), HybridDomain::Alpha(b)) => Some(a.cmp(b)),
+            (HybridDomain::Beta(a), HybridDomain::Beta(b)) => Some(a.cmp(b)),
+            _ => None,
+        }
+    }
+}
+
+impl<AD, BD> Ord for HybridDomain<AD, BD>
+where
+    AD: Domain,
+    BD: Domain,
+{
+    fn cmp(&self, other: &Self) -> Ordering {
+        match self.partial_cmp(other) {
+            Some(ordering) => ordering,
+            None => panic!("can't compare different variants of `HybridDomain`"),
+        }
+    }
+}
+
+impl<AD, BD> AsRef<[u8]> for HybridDomain<AD, BD>
+where
+    AD: Domain,
+    BD: Domain,
+{
+    fn as_ref(&self) -> &[u8] {
+        match self {
+            HybridDomain::Alpha(alpha) => alpha.as_ref(),
+            HybridDomain::Beta(beta) => beta.as_ref(),
+        }
+    }
+}
+
+impl<AD, BD> Element for HybridDomain<AD, BD>
+where
+    AD: Domain,
+    BD: Domain,
+{
+    fn byte_len() -> usize {
+        32
```

Comment:
Is the 32 a constant independent of the particular alpha/beta domains?

Reply:
Both `Blake2sDomain` and `PedersenDomain` are 32 bytes and are the only types that implement `Domain`, so we make the assumption that a `HybridDomain` will always be 32 bytes.

Reply:
I don't think this is a good assumption. We should probably put a check in places that both the domains have the same length, and then delegate to that.

Reply:
otherwise we should hardcode this much more strictly than just in this one place

Reply:
I realized (by chance) that we're already doing this:

https://github.com/filecoin-project/rust-fil-proofs/blob/b40d12ab3aa6ef38bc7b92e558837136a89bb221/storage-proofs/src/util.rs#L7

This assumption is hard-coded and implied in the entire MT build process, if this is indeed the case then I'm going to leverage it as another argument against the `Element`/`Domain` abstraction tied to the MT.